File: src/main/java/dev/vrba/slack/tenor/bot/service/storage/ResultsStorageService.java
package dev.vrba.slack.tenor.bot.service.storage;

import io.micronaut.core.annotation.NonNull;
import jakarta.inject.Singleton;

import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

@Singleton
public class ResultsStorageService {

    @NonNull
    private final Map<String, SearchResults> storage = new ConcurrentHashMap<>();

    @NonNull
    private final ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();

    @NonNull
    public Optional<SearchResults> get(@NonNull String id) {
        return Optional.ofNullable(storage.get(id));
    }

    @NonNull
    public SearchResults store(@NonNull String id, @NonNull SearchResults results) {
        storage.put(id, results);
        scheduler.schedule(() -> storage.remove(id), 30, TimeUnit.MINUTES);

        return results;

    }
}

File: src/main/kotlin/dev/vrba/slack/service/storage/ResultsStorageService.kt
package dev.vrba.slack.service.storage

import jakarta.inject.Singleton
import java.util.concurrent.ConcurrentHashMap

@Singleton
class ResultsStorageService {
    // TODO: Store this in Redis?
    private val storage: MutableMap<String, Results> = ConcurrentHashMap()

    suspend fun get(view: String): Results? = storage[view]

    suspend fun store(
        view: String,
        results: Results,
    ): Results {
        storage[view] = results
        return results
    }
}
